components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -0,0 +1,778 @@
+  if (!path) {
+    return absl::nullopt;
+  }
+
+  auto* token_data_as_dict =
+      pinned_assets_pref.FindDictByDottedPath(path.value());
+  if (!token_data_as_dict) {
+    return absl::nullopt;
+  }
+
+  auto* cids = token_data_as_dict->FindList(kAssetUrlListKey);
+  if (!cids) {
+    return absl::nullopt;
+  }
+
+  std::vector<std::string> result;
+  for (const base::Value& item : *cids) {
+    result.push_back(item.GetString());
+  }
+
+  return result;
+}
+
+mojom::TokenPinStatusPtr BraveWalletPinService::GetTokenStatus(
+    const absl::optional<std::string>& service,
+    const mojom::BlockchainTokenPtr& token) {
+  const base::Value::Dict& pinned_assets_pref =
+      prefs_->GetDict(kPinnedErc721Assets);
+
+  auto path = GetPath(service, token);
+  if (!path) {
+    return nullptr;
+  }
+
+  auto* token_data_as_dict =
+      pinned_assets_pref.FindDictByDottedPath(path.value());
+  if (!token_data_as_dict) {
+    return mojom::TokenPinStatus::New(
+        mojom::TokenPinStatusCode::STATUS_NOT_PINNED, nullptr, base::Time());
+  }
+
+  auto* status = token_data_as_dict->FindString(kAssetStatus);
+  if (!status) {
+    return mojom::TokenPinStatus::New(
+        mojom::TokenPinStatusCode::STATUS_NOT_PINNED, nullptr, base::Time());
+  }
+
+  auto pin_status = StringToStatus(*status).value_or(
+      mojom::TokenPinStatusCode::STATUS_NOT_PINNED);
+  base::Time validate_timestamp;
+  mojom::PinErrorPtr error;
+
+  {
+    auto* validate_timestamp_value =
+        token_data_as_dict->Find(kValidateTimestamp);
+    if (validate_timestamp_value) {
+      validate_timestamp =
+          base::ValueToTime(validate_timestamp_value).value_or(base::Time());
+    }
+
+    auto* error_dict = token_data_as_dict->FindDict(kError);
+    if (error_dict) {
+      auto* error_message = error_dict->FindString(kErrorMessage);
+      auto* error_code = error_dict->FindString(kErrorCode);
+      if (error_code && error_message) {
+        error = mojom::PinError::New(
+            StringToErrorCode(*error_code)
+                .value_or(mojom::WalletPinServiceErrorCode::ERR_PINNING_FAILED),
+            *error_message);
+      }
+    }
+  }
+
+  return mojom::TokenPinStatus::New(pin_status, std::move(error),
+                                    validate_timestamp);
+}
+
+absl::optional<base::Time> BraveWalletPinService::GetLastValidateTime(
+    const absl::optional<std::string>& service,
+    const mojom::BlockchainTokenPtr& token) {
+  const base::Value::Dict& pinned_assets_pref =
+      prefs_->GetDict(kPinnedErc721Assets);
+
+  auto path = GetPath(service, token);
+  if (!path) {
+    return absl::nullopt;
+  }
+
+  auto* token_data_as_dict =
+      pinned_assets_pref.FindDictByDottedPath(path.value());
+  if (!token_data_as_dict) {
+    return absl::nullopt;
+  }
+
+  auto* time = token_data_as_dict->Find(kValidateTimestamp);
+  return base::ValueToTime(time);
+}
+
+void BraveWalletPinService::FinishAddingWithResult(
```

Comment:
Why do we have most arguments not used? I also think having this function doesn't add too much benefits, it seems more clear to just see `std::move(callback).Run(result, std::move(error));` directly in those callers.